src/Microsoft.IdentityModel.Tokens/JsonWebKeySet.cs

```diff
@@ -122,68 +120,100 @@ public JsonWebKeySet(string json, JsonSerializerSettings jsonSerializerSettings)
+        {
+            try
+            {
+                var keys = new List<SecurityKey>();
+
+                foreach (var certString in jsonWebKey.X5c)
+                {
+                    var key = new X509SecurityKey(new X509Certificate2(Convert.FromBase64String(certString)))
                     {
-                        try
-                        {
-                            SecurityKey key =
-                                 new RsaSecurityKey
-                                 (
-                                    new RSAParameters
-                                    {
-                                        Exponent = Base64UrlEncoder.DecodeBytes(webKey.E),
-                                        Modulus = Base64UrlEncoder.DecodeBytes(webKey.N),
-                                    }
-
-                                );
-                            key.KeyId = webKey.Kid;
-                            keys.Add(key);
-                        }
-                        catch (CryptographicException ex)
-                        {
-                            throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX10801, webKey.E, webKey.N), ex));
-                        }
-                        catch (FormatException ex)
-                        {
-                            throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX10801, webKey.E, webKey.N), ex));
-                        }
-                    }
+                        KeyId = jsonWebKey.Kid
+                    };
+
+                    keys.Add(key);
                 }
+
+                return keys;
             }
+            catch (Exception ex)
+            {
+                throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX10802, jsonWebKey.X5c[0]), ex));
+            }
+        }
+
+        private SecurityKey CreateRsaSecurityKey(JsonWebKey jsonWebKey)
+        {
+            try
+            {
+                var rsaParams = new RSAParameters
+                {
+                    Exponent = Base64UrlEncoder.DecodeBytes(jsonWebKey.E),
+                    Modulus = Base64UrlEncoder.DecodeBytes(jsonWebKey.N),
+                };
 
-            return keys;
+                return new RsaSecurityKey(rsaParams)
+                {
+                    KeyId = jsonWebKey.Kid
+                };
+            }
+            catch (Exception ex)
+            {
+                throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX10801, jsonWebKey.E, jsonWebKey.N), ex));
+            }
+        }
+
+        private SecurityKey CreateECDsaSecurityKey(JsonWebKey jsonWebKey)
+        {
+            try
+            {
+                var ecdsaAdapter = new ECDsaAdapter();
+                var ecdsa = ecdsaAdapter.CreateECDsa(jsonWebKey, false);
```

Comment:
ECDsaAdapter.CreateECDsa is not static as we need to determine what is available on the machine. Could we create one adapter an reuse it?

Reply:
Is this meant to improve performance when we have more than one EC key present in JsonWebKeySet? If so, I agree.

Reply:
addressed with https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1143/commits/5f40af68917573f4fae7a80c0bb542f6ad275b5e